[amdgpu] Tuning the number of teams and threads

dhruvachak opened the pull request:
Tuning the number of teams and threads used for a kernel.  Includes a corner case fix.

ronlieb commented on openmp/libomptarget/plugins/amdgpu/src/rtl.cpp:
```diff
@@ -1767,10 +1805,10 @@ int32_t __tgt_rtl_run_target_team_region_locked(
             "DEVID:%2d SGN:%1d ConstWGSize:%-4d args:%2d teamsXthrds:(%4dX%4d) "
-            "reqd:(%4dX%4d) n:%s\n",
+            "reqd:(%4dX%4d) sgpr_count:%u vgpr_count:%u tripcount:%lu n:%s\n",
             device_id, KernelInfo->ExecutionMode, KernelInfo->ConstWGSize,
```
i like the addition of this info to the tracing.

ronlieb commented on openmp/libomptarget/plugins/amdgpu/impl/internal.h:
```diff
@@ -97,6 +97,8 @@ typedef struct atl_kernel_info_s {
   uint32_t private_segment_size;
+  uint32_t sgpr_count;
+  uint32_t vgpr_count;
   uint32_t kernel_segment_size;
```
could you also add sgpr_spill and vgpr_spill

dhruvachak replied:
Added spill counts to the next version.

gregrodgers reviewed (commented):
We need to start extending and making use of Grid values that are constant across compiler, host runtime and device runtime.   These are definined in OMPGridValues.h but the comments/documentation is wrong with respect to host runtime.   This is an example useage in the amdgpu plugin.   
 llvm::omp::AMDGPUGpuGridValues[llvm::omp::GVIDX::GV_Max_Teams];

I don't know if the deviceRTLs are using these enums to avoid duplicating these constants in macros. 

Greg

dhruvachak commented:
I added the vector_register_count to the grid values header. The other 2 macro defs that I introduced are really about the heuristics for determining the launch parameters and not directly related to the grid values. So I left them as macros. In some ways, they were already there in the code, hardcoded within a function def. I moved them to macros for better readability.

ronlieb commented:
after this lands in aomp13, it would be good to see if can be upstreamed. if all of it can, then good, it will land in amd-stg-open in due course. If not, let me know what part can, if any. so we can figure out how best to move it forward.

JonChesterfield commented on openmp/libomptarget/plugins/amdgpu/src/rtl.cpp:
```diff
@@ -48,6 +48,15 @@
 #define DEBUG_PREFIX "Target " GETNAME(TARGET_NAME) " RTL"
 
+// Total number of VGPRs per CU
+#define OMP_AMD_TOTAL_VGPR_PER_CU (64*1024)
```
trunk would want this spelled `static const unsigned ... = 64*1024`

JonChesterfield commented on openmp/libomptarget/plugins/amdgpu/src/rtl.cpp:
```diff
@@ -1748,13 +1757,42 @@ int32_t __tgt_rtl_run_target_team_region_locked(
+  else if (threadsPerGroup < RTLDeviceInfoTy::Warp_Size) {
+    // Lower bound is a wavefront size
+    threadsPerGroup = RTLDeviceInfoTy::Warp_Size;
+  }
```
this doesn't look clang-format'ed

JonChesterfield commented:
The authoritative copy of this library is the one in llvm upstream, please propose the patch against that instead and then we'll cherry-pick onto aomp13. Macros -> named constants and clang-format strongly recommended.

It would be good to split this into the change that records (and prints) sgpr/vgpr counts and a separate change that uses them to change the number of teams/threads. If the latter causes problems we may need to revert it, and splitting the patch means the *gpr change would survive said revert.

dhruvachak commented:
Merged.
